Peer-to-peer upgrades

hackergrrl opened the pull request:
Work-in-progress. The backend (`src/backend/`) is mostly finished & is ready for review.

gmaclennan committed:
Add AppInfo native module for accessing sourceDir

gmaclennan committed:
Add share sheet for sharing Mapeo APK

gmaclennan committed:
Test auto-updating app from an APK

gmaclennan committed:
Test reading package signatures

hackergrrl committed:
Prep APK for sharing and start upgrade server.

hackergrrl committed:
get correct mapeo app version + show on settings

hackergrrl committed:
remove old comments

hackergrrl committed:
implement upgrade storage module

hackergrrl committed:
prevent 'filename' from being exposed on upgrades

hackergrrl committed:
add createReadStream API to stream upgrade data

hackergrrl committed:
chore: add eslint linting

hackergrrl committed:
test: move fake.apk to static/ subdir

hackergrrl committed:
test: formatting

hackergrrl committed:
fix: make 'arch' an array of strings

hackergrrl committed:
chore: add 'test' npm script

hackergrrl committed:
add initial UpgradeServer implementation

hackergrrl committed:
add new UpgradeServer implementation

hackergrrl committed:
test: add edge case tests for upgrade server

hackergrrl committed:
move the upgrade discovery key to its own file

hackergrrl committed:
have UpgradeServer accept a port to advertise on

hackergrrl committed:
make UpgradeStorage#getAvailableUpgrades sync

hackergrrl committed:
refactor: consolidate apk->upgradeoption logic

hackergrrl committed:
Revert "make UpgradeStorage#getAvailableUpgrades sync"

This reverts commit 2b4e3ac7dca59da5e8690e3c3038d8301ef102e9.

hackergrrl committed:
refactor: use a file for downloaded upgrade info

hackergrrl committed:
refactor: move LocalUpgradeInfo into a nodule

hackergrrl committed:
refactor: rename createApkWriteStream

hackergrrl committed:
add work-in-progress UpgradeDownloader component

hackergrrl committed:
chore: lint

hackergrrl committed:
add download progress tracking to state

hackergrrl committed:
implement the UpgradeDownloader checker component

hackergrrl committed:
chore: cleanup

hackergrrl committed:
implement the UpgradeManager

hackergrrl committed:
fix: ignore non-mapeo app keys

hackergrrl committed:
chore: add eslintrc for backend

hackergrrl committed:
don't access UpgradeDownloader's internals

hackergrrl committed:
auto-download upgrades that UpgradeManager finds

hackergrrl committed:
chore: lint

hackergrrl committed:
UpgradeDownload doesn't need version/arch/etc args

hackergrrl committed:
fix bugs in case where server waits to shut down

hackergrrl committed:
upgrade-server runs its own http server

hackergrrl committed:
expose 'startTime' context for upgrade searches

hackergrrl committed:
chore: upgrade deps

hackergrrl committed:
p2p-upgrades: pass in version to storage

hackergrrl committed:
wip: integrate upgrademanager into server

hackergrrl committed:
chore: add missing dep from develop

hackergrrl committed:
workaround for github:digidem/mapeo-mobile#521

hackergrrl committed:
wip: fixes

hackergrrl committed:
test: lint

hackergrrl committed:
fix: be able to listen for multiple ready events

hackergrrl committed:
add upload info to sharing state

hackergrrl committed:
DotIndicator: accept optional size parameter

hackergrrl committed:
wip: frontend implementation

hackergrrl committed:
move upgrade state constants to a common module

hackergrrl committed:
backend: upgrade simulator tool

hackergrrl committed:
implement upgrade state validation module

hackergrrl committed:
use a common storage path for downloaded APKs

hackergrrl committed:
expose filename property from upgrade-storage api

hackergrrl committed:
clone upgrade state context when modifying

hackergrrl committed:
add more thorough upgrade state management

hackergrrl committed:
fix: expose upgrade filename instead of hash

hackergrrl committed:
add more thorough state mgmt to upgrade-server

hackergrrl committed:
add more upgrade state mgmnt to upgrade-manager

hackergrrl committed:
wip: variety of frontend implementation upgrade changes

hackergrrl committed:
prevent older APKs from trying to be installed

hackergrrl committed:
simplify frontend state handling of upgrades

hackergrrl committed:
add another effect to track peer + upgrade state

hackergrrl committed:
test: fixes

hackergrrl committed:
test: files aren't left over after write error

hackergrrl committed:
feat: check hash upon upgrade download + cleanup

hackergrrl committed:
storage: use a temp dir for in-progress downloads

hackergrrl committed:
p2p-upgrades: make logging prefixes consistent

hackergrrl committed:
lint

hackergrrl committed:
add debug logging to upgrade backend

hackergrrl committed:
fix: write completed upgrade to root storage dir

gmaclennan commented on android/app/src/main/java/com/mapeo/MainApplication.java:
```diff
@@ -46,16 +46,12 @@ public boolean getUseDeveloperSupport() {
-      );
-      packages.addAll(unimodules);
-
+      packages.add(new ModuleRegistryAdapter(mModuleRegistryProvider));
```
Does this actually work? My understanding was that this is a list of modules. These "unimodules" are the `expo-` modules that we use for:

- Camera
- Geolocation (GPS)
- Detecting device rotation for landscape photos

Can we check all of this still works in a build with this option?

hackergrrl replied:
`git blame` shows that this was code you wrote (maybe for the `share-apk` branch?). I don't know what any of this does :D

Results
- Camera: yes
- GPS: yes
- Orientation: yes

gmaclennan commented on android/app/src/main/java/com/mapeo/MainApplication.java:
```diff
@@ -75,37 +71,5 @@ public void onCreate() {
     super.onCreate();
     BugsnagReactNative.start(this);
     SoLoader.init(this, /* native exopackage */ false);
-    initializeFlipper(this, getReactNativeHost().getReactInstanceManager());
```
Any particular reason why we're removing Flipper here?

hackergrrl replied:
I think these java changes are all because of the drift from the original `share-apk` branch that this branch is based on, and `develop`. I'm going to to restore this code.

gmaclennan commented on src/backend/lib/constants.js:
```diff
@@ -0,0 +1,36 @@
+const UPGRADE_INFO_FILENAME = "upgrades.json";
+
+const UpgradeState = {
+  Search: {
+    Idle: 1,
+    Searching: 2,
+    Error: 3,
+  },
+
+  Download: {
+    Idle: 1,
+    Downloading: 2,
+    Downloaded: 3,
+    Error: 4,
+  },
+
+  Check: {
+    NotAvailable: 1,
+    Available: 2,
+    Error: 3,
+  },
+
+  Server: {
+    Idle: 1,
+    Sharing: 2,
+    Draining: 3,
+    Error: 4,
+  },
+};
```
This is a style preference, but I prefer strings for state, because it makes debugging easier. If I log the state value then I know what it is, but when it's a number I need to look it up.

gmaclennan replied:
PS. When I've been writing with Typescript/Flow, then I have dropped using constants, and instead just typed the state variable. Without Typescript I appreciate that using a constant avoids nasty typo errors!

hackergrrl replied:
Good idea! I'll change it over.

gmaclennan commented on src/backend/lib/local-upgrade-info.js:
```diff
@@ -0,0 +1,102 @@
+const fs = require("fs");
+const RWLock = require("rwlock");
+
+const UPGRADE_INFO_FILENAME = require("./constants").UPGRADE_INFO_FILENAME;
```
```suggestion
const { UPGRADE_INFO_FILENAME } = require("./constants");
```

Really not bothered about this, but I think most of the code in Mapeo uses destructuring for this kind of import.

gmaclennan commented on src/backend/lib/upgrade-manager.js:
```diff
@@ -0,0 +1,147 @@
+const validate = require("./validate-upgrade-state");
+
+class UpgradeManager {
+  constructor(dir, port, currentVersion, emitFn, listenFn, removeFn) {
```
```suggestion
  constructor({ dir, port, currentVersion, emitFn, listenFn, removeFn}) {
```

Code style pref: With this many arguments it is easy for a mistake to creep in when calling the constructor by missing an argument or getting them in the wrong order (I've spent some time debugging issues like this). Passing arguments as properties on an `options` object has less chance of error, and is easy with destructuring like this.

gmaclennan commented on src/backend/package.json:
```diff
@@ -9,29 +9,42 @@
     "@bugsnag/js": "^6.3.1",
     "bindings-noderify-nodejs-mobile": "^10.3.0",
-    "debug": "^4.1.1",
-    "discovery-swarm": "6.0.0",
```
I remember this was in here to lock the version of discovery-swarm used by mapeo-core, since we were seeing errors in ^6.0.0. Do you know if the original issue was resolved? I only have vague memories of this.

gmaclennan commented on src/backend/server.js:
```diff
@@ -131,10 +142,61 @@ function createServer({ privateStorage, sharedStorage, flavor }) {
+    getport((err, port) => {
+      if (err) return cb(err);
+
+      const emitFn = rnBridge.channel.post.bind(rnBridge.channel);
+      const listenFn = rnBridge.channel.on.bind(rnBridge.channel);
+      const removeFn = rnBridge.channel.removeListener.bind(rnBridge.channel);
+      upgradeManager = new UpgradeManager(
+        upgradePath,
+        port,
+        version,
+        emitFn,
+        listenFn,
+        removeFn
+      );
```
This architecture makes me nervous... I came here looking for the boundary between UpgradeManager and the `backend/server.js` code that hooks up the front-end. Giving UpgradeManager responsibility for managing listeners on the IPC channel feels prone to bugs, and figuring out exactly when and how events are emitted to the front-end is harder to follow. Is there a different way of implementing this?

hackergrrl replied:
What bugs is this prone to?

gmaclennan commented on src/frontend/api.js:
```diff
@@ -6,13 +6,18 @@ import nodejs from "nodejs-mobile-react-native";
   Messages,
 } from "./context/ConfigContext";
+
+import { version as APP_VERSION } from "../../package.json";
```
The actual version of Mapeo Mobile (returned by `react-native-device-info`) includes a build number, e.g. `2.4.1.839`. There are scenarios where we might have a broken build and release a new build that only increments the build number. Also, we _may_ change the build process where the version in `package.json` is not kept in sync, so I think better to read the version from `DeviceInfo` which will always be set.

gmaclennan commented on src/frontend/api.js:
```diff
@@ -137,6 +142,19 @@ export function Api({
 
   nodejs.channel.addListener("status", onStatus);
 
+  nodejs.channel.addListener("copy-apk", async () => {
+    const tmpDir = RNFS.DocumentDirectoryPath + "/installer";
+    log("+++ 5", tmpDir);
+    const apkName = "mapeo.apk";
+    const tmpApkPath = `${tmpDir}/${apkName}`;
+    await RNFS.mkdir(tmpDir);
+    log("+++ 6");
+    await RNFS.copyFile(AppInfo.sourceDir, tmpApkPath);
+    log("+++ 7");
+    // await RNFS.unlink(tmpDir);
+    nodejs.channel.post("apk-ready", tmpApkPath, APP_VERSION);
+  });
+
```
Why do the file copy in front-end code rather than nodejs-mobile? This feels like it belongs in the backend.

hackergrrl replied:
This is how it was done in the `share-apk` branch you wrote, so I kept it here, since it worked fine.

gmaclennan reviewed (commented):
Wow @noffle this is a massive amount of work! I only had 30 mins just now to do a very quick read through and leave initial comments. I was actually here reading through to answer https://github.com/digidem/mapeo-mobile/issues/528 and https://github.com/digidem/mapeo-mobile/issues/529, but it may be until next week that I can do a more thorough review of this, so I thought I would leave the thoughts/comments that came to me on the first read here as a starter. I'll answer your #528 #529 directly now that I have some understanding of the code.

gmaclennan committed:
Internationalization example

hackergrrl committed:
fix(upgrade): remove TTL; re-announce regularly

hackergrrl committed:
fix: readd flipper code (merge issue)

noffle committed:
Update src/backend/lib/local-upgrade-info.js

Co-authored-by: Gregor MacLennan <gmaclennan@digital-democracy.org>

hackergrrl reviewed (commented):
Looking over the full diff, I think there's a bunch of junk code from the `share-apk` branch I based this off of, so I appreciate the "why is this even here?" style comments :)

hackergrrl committed:
use strings for upgrade state constants

gmaclennan committed:
Internationalize text in update bar

gmaclennan committed:
Remove unused import

gmaclennan committed:
chore(linting): Combine duplicate imports

gmaclennan committed:
chore(linting): Remove unreachable code

gmaclennan committed:
Extract messages for translation

gmaclennan commented:
@noffle, sorry my review of this is coming in drips and drabs, but I wanted to get initial thoughts to you rather than delaying to get you everything at once.

As I mentioned on our call right now, I really like how the code is structured here, and that everything is essentially behind the `upgrade-manager.js`, i.e. there is a clear boundary between the upgrade code and the rest of the app. However, with the current patterns of passing `emitFn, listenFn, removeFn` to `UpgradeManager`, the API surface is not very clear.

I think it would be easier for others who develop the app if there is a clear public API to `UpgradeManager`, and we do the wiring up with the nodejs RPC outside it. That way also the implementation of `UpgradeManager` is not tied to how we have implemented communication between frontend-backend.

As I understand it, `UpgradeManager` provides 3 public methods:

1. `getState()`
2. `startServices()`
3. `stopServices()`

And it is an event emitter that emits two types of event:

1. `state`
2. `error`

I think we should write it as such, and then, outside of `upgrade-manager.js` in `src/backend/server.js` write the wiring up code, e.g. 

```js
rnBridge.channel.on('p2p-upgrade::get-state', () => {
  rnBridge.channel.post('p2p-upgrade::state', upgradeManager.getState())
})
upgradeManager.on('state', state => {
  rnBridge.channel.post('p2p-upgrade::state', state)
}
```

That way also the code in `server.js` is responsible for cleaning up these listeners on close (which it already does for other RPC listeners)

gmaclennan commented:
The other point I mentioned is the copying of the APK. This was only necessary for the testing that I did sharing the APK via the Android Share dialog - I needed to move the APK into a different part of Android storage so that I could add fileshare permissions (that is what the [`filepaths.xml`](https://github.com/digidem/mapeo-mobile/pull/524/files#diff-a029272cc71f7b06d507bc9b3507570edd16da59c9a85c0cac4fba592864b30c) code is about - it is letting Android know that we want the app to be able to share files in these folders with other apps via the Android Share screen - again this is not necessary now)

The nodejs process should be able to read the APK in its original location, so no need for the copy step. Only the frontend knows where this file is, but rather than have a back-and-forth when you start the upgrade manager, I suggest that we pass the APK path as part of the `config` rnBridge call:

https://github.com/digidem/mapeo-mobile/blob/share-apk-p2p/src/backend/index.js#L75
https://github.com/digidem/mapeo-mobile/blob/share-apk-p2p/src/frontend/api.js#L239

That way you can pass the APK path as a parameter when you first create a new `UpgradeManager` instance. I think you could remove the `apkReady` function in `server.js` if you did this.

hackergrrl commented:
This is great news -- it will simplify things considerably.

hackergrrl commented:
Agreed, I like this approach.

gmaclennan commented:
> Agreed, I like this approach.

Ok, great. In terms of the throttle of state events, I think then the place to put it is to throttle the function passed to the event handler `upgradeManager.on('state', throttledFunction)` -- I can do this once you change the API.

gmaclennan committed:
feat: Add minSdkVersion to AppInfo interface

gmaclennan committed:
chore: Add context to backend status errors

gmaclennan committed:
chore: Remove unused "flavor" option in createServer

gmaclennan committed:
chore: Don't try to support multiple "config" RPC messages

We wait for the config RPC message to start the server. This code would
try to restart the server with new config, but we don't have a use for
that and it makes the code a lot more complicated to manage, so I'm
removing it for now - it was not tested anyway and probably did not work

gmaclennan committed:
feat: Pass apkPath, minSdkVersion etc. to createServer()

apkPath: Path to currently running APK (file can be read from nodejs)
minSdkVersion: Min Android SDK version supported by currently running APK
version: Version name of running APK (semver if release, but not if testing)
buildNumber: Increments every time CI build happens - must always increase
bundleId: Different variants have different bundleIds
 - com.mapeo
 - com.mapeo.qa
 - com.mapeo.debug
 - com.mapeo.icca
Different bundle ids install as separate apps

hackergrrl committed:
java: add missing import

hackergrrl committed:
fix: prevent redownload of locally-stored upgrades

luandro committed:
chore: have UpgradeManager expose functions instead of events and ajust tests

hackergrrl committed:
do not copy the APK; use privateStorage instead

hackergrrl committed:
add optional param to upgrade simulator

gmaclennan committed:
chore: fix line-endings for gradlew.bat (should be CRLF)

gmaclennan committed:
Merge branch 'develop' into share-apk-p2p

* develop:
  feat: Setting to choose coordinate display format (#526)
  chore: Re-apply c267a99 with correct npm@6 version and lockfile v1
  Revert "feat: Updated translations for default configuration (vi, srn)"
  feat: Updated translations for default configuration (vi, srn)
  fix: Update Thai, Khmer & Vietnamese translations
  feat: Add Dutch translations
  feat: Add Sranan Tongo translations
  feat: Add French translations
  fix: Fix cutoff of text on OnePlus6T phone #502 (#511)
  fix: Write preset tags to observations (#514)
  chore: Fix Github CI builds by removing NDK r22 (#516)
  fix: Fix import config button crash (#512)
  feat: Updated translations (vi, es, po) (#498)

# Conflicts:
#	src/backend/package-lock.json
#	src/backend/package.json
#	src/frontend/screens/Settings/Settings.js

gmaclennan commented:
For discussion: How to handle different version numbers?

Mapeo version names are set at build time, and there are 3 types of releases we make:

1. **"Production" releases**: These have the format `{major}.{minor}.{patch}.{buildNumber}` where `buildNumber` is an integer that is guaranteed to increment for every build (including builds of the other "non-production" releases)
2. **Release Candidates**: Format `{major}.{minor}.{patch}-RC.{buildNumber}`. The `{major}.{minor}.{patch}` portion is the target production release version. Each subsequent build of an RC (e.g. for fixing bugs identified in QA) will keep the same `{major}.{minor}.{patch}` but `buildNumber` will increment.
3. **Internal test builds**: These are built for every PR against the `develop` branch, and they have the version format: `internal.{buildNumber}`. These are used to test a specific feature before a PR is merged.

I think the upgrade logic that makes most sense to me is:

1. Production releases only update to newer production releases. We could use _only_ the build number for this (since it is guaranteed to increment for every new production release), but in the future we might want to use the major/minor/patch in order to have different behaviour based on that (e.g. we might treat a major update differently from a minor in the UX)
2. Release candidate releases should update to newer RCs _for the same `{major}.{minor}.{patch}`_, but if the major/minor/patch changes they should be ignored.
3. Internal test builds should update to a newer internal test build from the same PR, based on the incrementing build number.

Since these version numbers are set as part of our CI build, when testing local builds, the version name will be `0.0.0` by default. In order to build locally with a different version name, set the `ANDROID_VERSION_NAME` env variable, e.g. `ANDROID_VERSION_NAME=1.1.0.0 npm run android` will build an APK where `DeviceInfo.version()` returns `1.1.0.0`.

I think the current version naming scheme is confusing and hard to parse (due to not being clearly defined). I propose that we switch to semver compatible versions, see https://github.com/digidem/mapeo-mobile/issues/554

hackergrrl commented:
Sounds good to me!

gmaclennan commented:
@noffle Can we remove this commit from here: https://github.com/digidem/mapeo-mobile/commit/28271b1a0166228bc05847208247bde55d3ccec7 This was just an experiment and the code is unfinished (I was trying to learn how to check signatures on the APK, but installing an incorrectly signed APK will not work anyway).

hackergrrl committed:
Revert "Test reading package signatures"

This reverts commit 28271b1a0166228bc05847208247bde55d3ccec7.

hackergrrl commented:
@gmaclennan sure! pushed a revert commit!

noffle committed:
Remove redundant keys from backend setup

Co-authored-by: Gregor MacLennan <gmaclennan@digital-democracy.org>

noffle committed:
Merge pull request #566 from digidem/share-apk-p2p-gregor2

Add additional config available for UpgradeManager instance

hackergrrl committed:
remove unused error handling

gmaclennan committed:
Fix naming of apkVersion

gmaclennan committed:
Merge branch 'share-apk-p2p' into issue-536-2

* share-apk-p2p:
  Fix naming of apkVersion
  remove unused error handling
  Remove redundant keys from backend setup
  Revert "Test reading package signatures"
  feat: Pass apkPath, minSdkVersion etc. to createServer()
  chore: Don't try to support multiple "config" RPC messages
  chore: Remove unused "flavor" option in createServer
  chore: Add context to backend status errors
  feat: Add minSdkVersion to AppInfo interface

luandro committed:
Opt-in setting for p2p upgrades (#561)

* chore: cleanup old items in settings

* chore: add messages for experiments

* feat: Add Experiments page, route and state to SettingsContext

* feat: Only start p2p-upgrade if enabled on settings

* chore: Remove redundant code

* Add experimental flag to useEffect deps

Without this the useEffect will not run if this value changes

* fix: boolean type is lowercase

* Change shape of settings context

We plan to use this for other features in the future, so adding
`experiments` as a property on settingsContext.

* Set test IDs

Co-authored-by: Gregor MacLennan <gmaclennan@digital-democracy.org>

gmaclennan committed:
fix: Don't copy src/backend/test/fake.apk as part of CI Gradle build

gmaclennan committed:
Update backend build script to only copy whitelisted files

The backend build script uses noderify to bundle JS into a single file.
Previously the script would then blacklist unnecessary files to delete
them from the packaged app, but that is a maintenance burden.
This change whitelists the necessary files that must be included in the
APK and only includes those.

hackergrrl committed:
fix: use correct variable names

gmaclennan committed:
chore: Run backend tests on CI (#577)

* chore: Run backend tests on CI

* chore: Run backend tests in parallel with e2e tests

gmaclennan committed:
fix(AppInfo): Remove minSdkVersion

This property is only available in Android API 24 and later (Android v7)
We planned to use it to check compatibility incase our supported
minSdkVersion increased (it is currently API 20). Since this is not
available in phones we support, it is not useful for this purpose, and
using it causes a crash on phones running Android <v7

gmaclennan committed:
Revert "fix: use correct variable names"

This reverts commit 4526fba8e095ec1dc3e4abaa20b4310b884401e9.
Variable names were already corrected in 5709722a2fab65ab1fb330bd9566ca66bbfe5c3e

gmaclennan committed:
Merge branch 'share-apk-p2p' into issue-536-2

* share-apk-p2p:
  Revert "fix: use correct variable names"
  fix(AppInfo): Remove minSdkVersion
  chore: Run backend tests on CI (#577)
  fix: use correct variable names
  Update backend build script to only copy whitelisted files
  fix: Don't copy src/backend/test/fake.apk as part of CI Gradle build
  Opt-in setting for p2p upgrades (#561)

gmaclennan committed:
Add integration with server.js

gmaclennan committed:
Fix upgrade simulator for new upgrademanager pattern

gmaclennan committed:
Update settings store key for new settings shape

This will cause anyone who has used the "Coordinate format" setting to
loose their chosen format, but we have not released that yet. For the
next time we update the settings that are persisted, we need a way to
migrate old settings, because leaving this at @1 broke my phone, because
it had stored settings that did not have the experiments option.

gmaclennan committed:
Remove changes that should not be in this PR

gmaclennan committed:
Merge branch 'share-apk-p2p' into issue-536-2

* share-apk-p2p:
  Update settings store key for new settings shape

hackergrrl committed:
fix: upgrade semver sorting

hackergrrl committed:
fix: add upgrade options listener only once

hackergrrl committed:
fix: check for upgrades after current d/l finishes

gmaclennan commented on src/backend/lib/upgrade-server.js:
```diff
@@ -0,0 +1,271 @@
+        clearInterval(this.announceInterval);
+        this.announceInterval = null;
+
+        // XXX: 'unannounce' seems to always return an error here; ignoring.
```
Yes, I checked the dns-discovery code, and there is a bug where this will always return an error (because `success` is never set to true on the unannounce path when no servers are defined) and will never emit `unannounce`. https://github.com/mafintosh/dns-discovery/blob/master/index.js#L425-L482

gmaclennan commented on src/backend/lib/upgrade-server.js:
```diff
@@ -0,0 +1,271 @@
+        this.discovery.unannounce(DISCOVERY_KEY, this.port, _ => {
+          log("drain: ..discovery unannounced");
+          this.discovery.destroy(err => {
+            if (err) {
```
The destroy callback never has any arguments (`err` will always be undefined) see https://github.com/mafintosh/dns-discovery/blob/master/index.js#L605-L625

hackergrrl committed:
fix: prioritize Downloading over ReadyToUpgrade

hackergrrl committed:
fix: reset upgrade state on download completion

hackergrrl committed:
fix: remove extra setState; fix param on a setState

hackergrrl committed:
fix: remove setState(IDLE) on error

hackergrrl committed:
test: fixes

These are some fixes re: recent changes to the upgrade state machines
that needed correction in the tests. These fixes are pretty hacky, but I
didn't feel it made sense to invest time into rewriting them right now.

hackergrrl committed:
fix: don't show search:IDLE as an error state

This is normal behaviour: the frontend will generally do a 'get-state'
request before search has initialized, revealing it as IDLE.

gmaclennan committed:
chore: Fix backend tests CI

gmaclennan commented on src/backend/lib/upgrade-download.js:
```diff
@@ -0,0 +1,303 @@
+          filename,
+          option.version,
+          option.hash,
+          err => {
+            if (err) this.setState(UpgradeState.Download.Error, err);
+            else this.setState(UpgradeState.Download.Downloaded, null);
+          }
```
It doesn't seem to make sense to mix a stream and a callback, because we end up with two ways this completes: when the `pump` pipeline completes, and then when this callback is called, which happen at different times. I feel like `createApkWriteStream` should return a writable stream which will `finish` when the APK is written.

gmaclennan reviewed (commented):
I've had a chance to more thoroughly review the code in this PR. It took me a while because it is so large and complex.

### State modelling

I think some states are not being modelled, which is a problem because the behaviour in these states is hard to understand from the code and can result in unexpected behavior when things happen suring this missing states.

Specifically: I can see 4 states in the `Search` module:

1. `IDLE`
2. `STARTING`
3. `SEARCHING`
4. `STOPPING`

(and maybe `ERROR` if it is possible for an un-recoverable error to happen)

Currently the code works around not being explicit about these states by using the `rwlock` module, however this makes it unclear about what will happen as a result of any action in these states. E.g. what does the `onPeer` event do if the state is in `STARTING` or `STOPPING`. Right now that requires understanding how the `rwlock` module works and the rules that it follows (parallel reads, but only after writes are complete) and requires knowing the exact order of the history e.g. what happens when onPeer is called after reset is called but before it has started?

Similarly the state model for the `Server` class is missing the `STARTING` state. Again the work-around using `rwlock` makes it hard to reason about what is happening and catch bugs.

I think using `rwlock` as a way to update state makes it hard to reason with the code and introduces the possibility of bugs, due to the difficulty to model what happens in the transition states. It essentially queues up state updates and reads, but in a way that is not that obvious and might not be what we actually want.

Can we explicitly model these transition states, and then we can be explicit in the other methods (e.g. `onPeer` or the server methods) about what happens in these states.

### Version comparison

There are a few places with logic to compare versions: in `upgrade-download.js`, `upgrade-storage.js` and in `upgrade-manager.js` (although this is just a semver compare here). Can we use a single function for comparing versions everywhere, otherwise I see this as being easy to miss if we want to change the comparison logic in the future.

Also the current logic in the `Check.check()` method seems to just randomly choose an available option from storage, rather than the most recent (it filters the available upgrades then picks the one at index `0`).

### Storage and checks

It might be useful to make the `Storage` class stateful, so that it can maintain a state about the available upgrades to avoid the need for the `Check` class.

E.g. if `Storage` had a `availableUpgrades` state, that was emitted every time that it changes e.g.

- After initial startup and read of the disk
- When `createApkWriteStream` successfully completes

This way the upgrade manager would only need to listen for changes in the storage state to know when an upgrade is available, and it would not need the coordination between the `Check` module and the `Storage` module.

hackergrrl committed:
fix: remove DOWNLOADED upgrade state

hackergrrl committed:
clean up UpgradeManager download logic

hackergrrl committed:
remove unused code

hackergrrl committed:
allow UpgradeCheck to re-check + choose latest version

gmaclennan committed:
fix: Throttle upload/download progress events to 100ms

noffle committed:
Merge pull request #580 from digidem/share-apk-p2p-redownload

Support downloading subsequent newer versions of the app

gmaclennan committed:
Merge commit '49b7464e2ab170939139f32a1026e037f4edf7b9' into issue-536-2

* commit '49b7464e2ab170939139f32a1026e037f4edf7b9':
  allow UpgradeCheck to re-check + choose latest version
  remove unused code
  clean up UpgradeManager download logic
  fix: remove DOWNLOADED upgrade state
  fix: don't show search:IDLE as an error state
  test: fixes
  fix: remove setState(IDLE) on error
  fix: remove extra setState; fix param on a setState
  fix: reset upgrade state on download completion
  fix: prioritize Downloading over ReadyToUpgrade
  fix: check for upgrades after current d/l finishes
  fix: add upgrade options listener only once
  fix: upgrade semver sorting
  chore: Fix backend tests CI

# Conflicts:
#	src/backend/lib/upgrade-manager.js

gmaclennan committed:
fix: Fix tests (maybe?)

gmaclennan committed:
Fix lint issues

gmaclennan committed:
fix: Throttle progress events to every 400ms

gmaclennan committed:
fix: Fix ApkInstaller for Android <7 (SDK version <24) (#586)

noffle committed:
Merge pull request #567 from digidem/issue-536-2

UpgradeManager expose functions

hackergrrl committed:
s/PROGRESS_THROTTLE/PROGRESS_THROTTLE_MS

noffle committed:
Merge pull request #588 from digidem/p2p-update/fix-progress-throttle

fix: Throttle upload/download progress events to 400ms

hackergrrl committed:
remove obsolete comment

hackergrrl committed:
Merge branch 'develop' into share-apk-p2p

gmaclennan commented:
Superseded by #595
